python-avd/pyavd/schema/metaschema/meta_schema_model.py

```diff
@@ -0,0 +1,559 @@
+
+
+class AvdSchemaBaseModel(BaseModel, ABC):
+    """
+    Base class for AvdSchema fields.
+
+    Contains nested models and common fields that applies to all subclasses.
+
+    Also covers internal attributes and methods used for documentation generation.
+    All these are prefixed with underscore.
+    """
+
+    # Common nested models used by common fields
+    class Deprecation(BaseModel):
+        """Deprecation settings"""
+
+        warning: bool = True
+        """Emit deprecation warning if key is set."""
+        new_key: str | None = None
+        """Relative path to new key."""
+        removed: bool | None = False
+        """Support for this key has been removed."""
+        remove_in_version: str | None = None
+        """Version in which the key will be removed."""
+        remove_after_date: str | None = None
+        """Date after which the key will be removed in the next major version."""
+        url: str | None = None
+        """URL detailing the deprecation and migration guidelines."""
+
+    # Pydantic config option to forbid keys in the inputs that are not covered by the model
+    model_config = ConfigDict(extra="forbid")
+
+    # Common field properties
+    display_name: str | None = Field(None, pattern=r"^[^\n]+$")
+    """Free text display name for forms and documentation (single line)"""
+    description: Annotated[str, constr(min_length=1)] | None = None
+    """Free text description for forms and documentation (multi line)"""
+    required: bool | None = None
+    """Key is required"""
+    deprecation: Deprecation | None = None
+    field_ref: str | None = Field(None, alias="$ref")
+    """
+    Reference to Sub Schema using JSON Path.
+    Example 'eos_designs#/keys/mykey' will resolve the schema for 'mykey' under the root dictionary of the eos_designs schema.
+    """
+    documentation_options: DocumentationOptions | None = None
+    """Schema field options used for controlling documentation generation"""
+
+    # Type of schema docs generators to use for this schema field.
+    _table_row_generator: type[TableRowGenBase]
+    _yaml_line_generator: type[YamlLineGenBase]
+
+    # Internal attributes used by schema docs generators
+    _key: str | None = None
+    _parent_schema: AvdSchemaField | None = None
+    _is_primary_key: bool = False
+    _is_first_list_key: bool = False
+
+    # Signal to __init__ if the $ref in the schema should be resolved before initilizing the pydantic model.
+    _resolve_schema: ClassVar[bool] = True
+
+    def __init__(self, resolve_schema: bool | None = None, **data):
+        """
+        Overrides BaseModel.__init__(**data).
+        Takes a kwarg "resolve_schema" which controls if all subclasses of AvdSchemaBaseModel should expand any $ref in the input schema.
+
+        The $ref expansion _only_ covers this field.
+        Any $ref on child fields are expanded as they are initialized by Pydantic since they are based on this base class.
```

Comment:
Is this still correct? I thought we discussed (and it seems to be implemented) the class attribute to be applied even on children?
It seems to contradict the line 112 comment

Reply:
There is a difference between child fields and subclasses.